src/common/crypto/platform/openssl/crypto.cpp

```diff
@@ -478,21 +478,61 @@ expected::ExpectedString ExtractPublicKey(const Args &args) {
 	if (ret != OPENSSL_SUCCESS) {
 		return expected::unexpected(MakeError(
 			SetupError,
-			"Failed to extract the public key from: (" + args.private_key_path
+			"Failed to extract the public key from the private key (" + args.private_key_path
 				+ "): OpenSSL BIO write failed: " + GetOpenSSLErrorMessage()));
 	}
 
-	int pending = BIO_ctrl_pending(bio_public_key.get());
+	// NOTE: At this point we already have a public key available for extraction.
+	// However, when using some providers in OpenSSL3 the external provider might
+	// write the key in the old PKCS#1 format. The format is not deprecated, but
+	// our older backends only understand the format if it is in the PKCS#8
+	// (SubjectPublicKey) format:
+	//
+	// For us who don't speak OpenSSL:
+	//
+	// -- BEGIN RSA PUBLIC KEY -- <- PKCS#1 (old format)
+	// -- BEGIN PUBLIC KEY -- <- PKCS#8 (new format - can hold different key types)
+
+
+	auto evp_public_key = PkeyPtr(
+		PEM_read_bio_PUBKEY(bio_public_key.get(), nullptr, nullptr, nullptr), pkey_free_func);
+
+	if (evp_public_key == nullptr) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to extract the public key from the private key " + args.private_key_path
+				+ "):" + GetOpenSSLErrorMessage()));
+	}
+
+	auto bio_public_key_new =
+		unique_ptr<BIO, void (*)(BIO *)>(BIO_new(BIO_s_mem()), bio_free_all_func);
+
+	if (bio_public_key_new == nullptr) {
+		return expected::unexpected(MakeError(
+			SetupError,
+			"Failed to extract the public key from the public key " + args.private_key_path
```

Comment:
This line still says "from the public key" + args.private_key

Reply:
Yes, I unified them to this. I couldn't really come up with something better. The error shows up in the error message, so it felt most right this way. But open to suggestions